File: swift/Sources/FieldUICore/Forces/CoreForces.swift
import Foundation
import simd

// MARK: - The canonical nine forces (§6)
//
// Each is an independent `Force` — the engine never changes to add one (§4). The math is
// the exact per-frame implementation from forces-system.md §6.1–§6.9, lifted to 3D:
// `env.vector` points from the particle toward the body; `env.dist ≥ 1`. Tangential terms
// use cross products against the plane axis (0,0,1) — identical to the JS 2D math at z=0,
// meaningful swirl-about-axis in a volume. On-state (`isEngaged`) widens range and boosts
// strength per the spec.

/// The out-of-plane axis 2D tangential math is defined against.
@usableFromInline let PLANE_AXIS = Vec3(0, 0, 1)

/// §6.1 — a soft gravity-like well, with optional orbital swirl.
public struct AttractForce: Force {
    public let token = "attract"
    public let label = "Attract"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.5 : 1)
        let s = b.strength * (b.isEngaged ? 3 : 1)
        if e.dist >= range { return }
        let f = pow(1 - e.dist / range, 2) * s * 0.5
        let u = e.vector / e.dist
        p.velocity += u * f
        if e.form.orbit != 0 {
            p.velocity += simd_cross(PLANE_AXIS, u) * (f * e.form.orbit) // tangential swirl → orbits
        }
        if b.isEngaged { p.heat = max(p.heat, (1 - e.dist / range) * 0.9) }
    }
}

/// §6.6 — inverse-square outward push; carves a void.
public struct RepelForce: Force {
    public let token = "repel"
    public let label = "Repel"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.4 : 1)
        let s = b.strength * (b.isEngaged ? 2 : 1)
        if e.dist >= range { return }
        let f = pow(1 - e.dist / range, 2) * s * 0.5
        p.velocity -= (e.vector / e.dist) * f
    }
}

/// §6.8 — tangential swirl with light inward retention.
public struct SwirlForce: Force {
    public let token = "swirl"
    public let label = "Swirl"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.4 : 1)
        let s = b.strength * (b.isEngaged ? 2 : 1)
        if e.dist >= range { return }
        let f = pow(1 - e.dist / range, 1.4) * s * 0.45
        let u = e.vector / e.dist
        // tangential swirl with a light inward retention (0.12): the swirl dominates ~8×,
        // so canonical swirl reads as a designed spin, not a drain.
        p.velocity += simd_cross(u, PLANE_AXIS) * (f * b.spin) + u * (f * 0.12)
        if b.isEngaged { p.heat = max(p.heat, (1 - e.dist / range) * 0.6) }
    }
}

/// §6.5 — a steady directional current along the heading.
public struct StreamForce: Force {
    public let token = "stream"
    public let label = "Stream"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.4 : 1)
        let s = b.strength * (b.isEngaged ? 2 : 1)
        if e.dist >= range { return }
        let f = pow(1 - e.dist / range, 1.1) * s * 0.5
        p.velocity += b.heading * f
        if b.isEngaged { p.heat = max(p.heat, (1 - e.dist / range) * 0.5) }
    }
}

/// §6.7 — viscosity; bleeds momentum, no redirection.
public struct ViscosityForce: Force {
    public let token = "viscosity"
    public let label = "Viscosity"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.4 : 1)
        if e.dist >= range { return }
        let k = (1 - e.dist / range) * (0.05 + b.strength * 0.07) * (b.isEngaged ? 1.6 : 1)
        p.velocity -= p.velocity * k
    }
}

/// §6.2 — a conduit: draws matter in, jets it out along the heading.
public struct JetForce: Force {
    public let token = "jet"
    public let label = "Jet"
    public let isKinematic = true // relaunches matter at the nozzle, so mass must not scale it
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let range = b.range * (b.isEngaged ? 1.4 : 1)
        if e.dist >= range { return }
        if e.dist < 24 {
            // at the nozzle: relaunch as a hot jet, with a cone of spread (about the plane axis).
            let sp = (Float.random(in: 0..<1) - 0.5) * 0.8
            let rot = simd_quatf(angle: sp, axis: PLANE_AXIS)
            let h = rot.act(b.heading)
            let spd = 2.4 + b.strength * 2.6
            p.velocity = h * spd
            p.position = b.center + h * 26
            p.heat = max(p.heat, 0.9)
        } else {
            // feed: draw surrounding matter toward the nozzle.
            let f = pow(1 - e.dist / range, 2) * (0.25 + b.strength * 0.15)
            p.velocity += (e.vector / e.dist) * f
        }
    }
}

/// §6.3 — a tether with a rest length; holds matter at a shell radius.
public struct TetherForce: Force {
    public let token = "tether"
    public let label = "Tether"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let rest = b.range * 0.6 * (b.isEngaged ? 1.25 : 1)
        let reach = rest * 2.1
        if e.dist >= reach { return }
        let k = (0.006 + b.strength * 0.012) * (b.isEngaged ? 1.7 : 1)
        let stretch = e.dist - rest
        p.velocity += (e.vector / e.dist) * (stretch * k)
        p.velocity *= 0.985
        if b.isEngaged {
            p.heat = max(p.heat, (1 - min(1, abs(stretch) / rest)) * 0.5)
        }
    }
}

/// §6.4 — an axis-aligned bouncing box; sparks on hard impact.
/// In 3D the reflection axis is the one of least penetration; a flat box (hd = 0)
/// only ever reflects x or y — exactly the JS wall.
public struct WallForce: Force {
    public let token = "wall"
    public let label = "Wall"
    public let isKinematic = true // an elastic bounce reflects velocity regardless of inertia
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        let pad: Float = 6
        let o = simd_abs(p.position - b.center)
        let he = b.box.halfExtents
        if o.x >= he.x + pad || o.y >= he.y + pad { return }
        let solid3D = he.z > 0
        if solid3D && o.z >= he.z + pad { return }
        let speed = simd_length(p.velocity)
        let px = he.x + pad - o.x
        let py = he.y + pad - o.y
        let pz = solid3D ? he.z + pad - o.z : Float.infinity
        if px < py && px < pz {
            p.position.x = p.position.x < b.center.x ? b.center.x - he.x - pad : b.center.x + he.x + pad
            p.velocity.x = -p.velocity.x * 0.85
        } else if py < pz {
            p.position.y = p.position.y < b.center.y ? b.center.y - he.y - pad : b.center.y + he.y + pad
            p.velocity.y = -p.velocity.y * 0.85
        } else {
            p.position.z = p.position.z < b.center.z ? b.center.z - he.z - pad : b.center.z + he.z + pad
            p.velocity.z = -p.velocity.z * 0.85
        }
        if speed > 0.7 {
            e.spark(p.position, min(2.4, speed), CANONICAL_FORCE_COLORS["wall"]) // canon spark tint (forces.config)
            p.heat = max(p.heat, min(0.85, speed * 0.4))
        }
    }
}

/// §6.9 — captures matter (held, conserved), then releases on saturation.
public struct SinkForce: Force {
    public let token = "sink"
    public let label = "Sink"
    public init() {}

    public func apply(body b: Body, particle p: Particle, env e: Env) {
        if p.cap != nil || e.dist >= b.absorbR { return }
        p.cap = b
        b.accreted += 1
        if b.accreted >= b.capacity { e.supernova(b) }
    }
}

/// The canonical nine, in spec order.
public func coreForces() -> [any Force] {
    [AttractForce(), JetForce(), TetherForce(), WallForce(), StreamForce(),
     RepelForce(), ViscosityForce(), SwirlForce(), SinkForce()]
}

File: android/fundamental-core/src/main/kotlin/com/fundamental/core/forces/CoreForces.kt
package com.fundamental.core.forces

import com.fundamental.core.engine.Body
import com.fundamental.core.engine.CANONICAL_FORCE_COLORS
import com.fundamental.core.engine.Env
import com.fundamental.core.engine.Force
import com.fundamental.core.engine.Particle
import com.fundamental.core.math.Vec3
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.pow

// The canonical nine forces (§6) — the Kotlin port of
// swift/Sources/FundamentalCore/Forces/CoreForces.swift. `env.vector` points from the particle toward
// the body; `env.dist ≥ 1`. Tangential terms use cross products against the plane axis (0,0,1) —
// identical to the JS 2D math at z = 0. On-state (`isEngaged`) widens range and boosts strength.
//
// Six of the nine (attract, repel, swirl, stream, tether, viscosity) are deterministic and held to
// the cross-plane conformance golden. The other three — jet (RNG cone), wall (box reflection), sink
// (stateful capture) — are not golden-testable by a single apply, so (exactly as the Swift port does)
// they are verified by behavioral unit tests instead. See CoreForcesBehaviorTests.

/** The out-of-plane axis the 2D tangential math is defined against. */
val PLANE_AXIS = Vec3(0f, 0f, 1f)

internal fun absVec(v: Vec3): Vec3 = Vec3(abs(v.x), abs(v.y), abs(v.z))

/** §6.1 — a soft gravity-like well, with optional orbital swirl. */
class AttractForce : Force {
    override val token = "attract"
    override val label = "Attract"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.5f else 1f)
        val s = body.strength * (if (body.isEngaged) 3f else 1f)
        if (env.dist >= range) return
        val f = (1f - env.dist / range).pow(2f) * s * 0.5f
        val u = env.vector / env.dist
        particle.velocity += u * f
        if (env.form.orbit != 0f) {
            particle.velocity += PLANE_AXIS.cross(u) * (f * env.form.orbit) // tangential swirl → orbits
        }
        if (body.isEngaged) particle.heat = maxOf(particle.heat, (1f - env.dist / range) * 0.9f)
    }
}

/** §6.6 — inverse-square outward push; carves a void. */
class RepelForce : Force {
    override val token = "repel"
    override val label = "Repel"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.4f else 1f)
        val s = body.strength * (if (body.isEngaged) 2f else 1f)
        if (env.dist >= range) return
        val f = (1f - env.dist / range).pow(2f) * s * 0.5f
        particle.velocity -= (env.vector / env.dist) * f
    }
}

/** §6.8 — tangential swirl with light inward retention. */
class SwirlForce : Force {
    override val token = "swirl"
    override val label = "Swirl"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.4f else 1f)
        val s = body.strength * (if (body.isEngaged) 2f else 1f)
        if (env.dist >= range) return
        val f = (1f - env.dist / range).pow(1.4f) * s * 0.45f
        val u = env.vector / env.dist
        // tangential swirl with a light inward retention (0.12): the swirl dominates ~8×, so
        // canonical swirl reads as a designed spin, not a drain.
        particle.velocity += u.cross(PLANE_AXIS) * (f * body.spin) + u * (f * 0.12f)
        if (body.isEngaged) particle.heat = maxOf(particle.heat, (1f - env.dist / range) * 0.6f)
    }
}

/** §6.5 — a steady directional current along the heading. */
class StreamForce : Force {
    override val token = "stream"
    override val label = "Stream"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.4f else 1f)
        val s = body.strength * (if (body.isEngaged) 2f else 1f)
        if (env.dist >= range) return
        val f = (1f - env.dist / range).pow(1.1f) * s * 0.5f
        particle.velocity += body.heading * f
        if (body.isEngaged) particle.heat = maxOf(particle.heat, (1f - env.dist / range) * 0.5f)
    }
}

/** §6.7 — viscosity; bleeds momentum, no redirection. */
class ViscosityForce : Force {
    override val token = "viscosity"
    override val label = "Viscosity"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.4f else 1f)
        if (env.dist >= range) return
        val k = (1f - env.dist / range) * (0.05f + body.strength * 0.07f) * (if (body.isEngaged) 1.6f else 1f)
        particle.velocity -= particle.velocity * k
    }
}

/** §6.3 — a tether with a rest length; holds matter at a shell radius. */
class TetherForce : Force {
    override val token = "tether"
    override val label = "Tether"

    override fun apply(body: Body, particle: Particle, env: Env) {
        val rest = body.range * 0.6f * (if (body.isEngaged) 1.25f else 1f)
        val reach = rest * 2.1f
        if (env.dist >= reach) return
        val k = (0.006f + body.strength * 0.012f) * (if (body.isEngaged) 1.7f else 1f)
        val stretch = env.dist - rest
        particle.velocity += (env.vector / env.dist) * (stretch * k)
        particle.velocity *= 0.985f
        if (body.isEngaged) {
            particle.heat = maxOf(particle.heat, (1f - min(1f, abs(stretch) / rest)) * 0.5f)
        }
    }
}

/** §6.2 — a conduit: draws matter in, jets it out along the heading. */
class JetForce : Force {
    override val token = "jet"
    override val label = "Jet"
    override val isKinematic = true // relaunches matter at the nozzle, so mass must not scale it

    override fun apply(body: Body, particle: Particle, env: Env) {
        val range = body.range * (if (body.isEngaged) 1.4f else 1f)
        if (env.dist >= range) return
        if (env.dist < 24f) {
            // at the nozzle: relaunch as a hot jet, with a cone of spread (about the plane axis).
            val sp = (env.rng() - 0.5f) * 0.8f
            val h = body.heading.rotatedAboutZ(sp)
            val spd = 2.4f + body.strength * 2.6f
            particle.velocity = h * spd
            particle.position = body.center + h * 26f
            particle.heat = maxOf(particle.heat, 0.9f)
        } else {
            // feed: draw surrounding matter toward the nozzle.
            val f = (1f - env.dist / range).pow(2f) * (0.25f + body.strength * 0.15f)
            particle.velocity += (env.vector / env.dist) * f
        }
    }
}

/**
 * §6.4 — an axis-aligned bouncing box; sparks on hard impact. In 3D the reflection axis is the one of
 * least penetration; a flat box (he.z == 0) only ever reflects x or y — exactly the JS wall.
 */
class WallForce : Force {
    override val token = "wall"
    override val label = "Wall"
    override val isKinematic = true // an elastic bounce reflects velocity regardless of inertia

    override fun apply(body: Body, particle: Particle, env: Env) {
        val pad = 6f
        val o = absVec(particle.position - body.center)
        val he = body.box.halfExtents
        if (o.x >= he.x + pad || o.y >= he.y + pad) return
        val solid3D = he.z > 0f
        if (solid3D && o.z >= he.z + pad) return
        val speed = particle.velocity.length()
        val px = he.x + pad - o.x
        val py = he.y + pad - o.y
        val pz = if (solid3D) he.z + pad - o.z else Float.POSITIVE_INFINITY
        val p = particle.position
        val v = particle.velocity
        if (px < py && px < pz) {
            val nx = if (p.x < body.center.x) body.center.x - he.x - pad else body.center.x + he.x + pad
            particle.position = Vec3(nx, p.y, p.z)
            particle.velocity = Vec3(-v.x * 0.85f, v.y, v.z)
        } else if (py < pz) {
            val ny = if (p.y < body.center.y) body.center.y - he.y - pad else body.center.y + he.y + pad
            particle.position = Vec3(p.x, ny, p.z)
            particle.velocity = Vec3(v.x, -v.y * 0.85f, v.z)
        } else {
            val nz = if (p.z < body.center.z) body.center.z - he.z - pad else body.center.z + he.z + pad
            particle.position = Vec3(p.x, p.y, nz)
            particle.velocity = Vec3(v.x, v.y, -v.z * 0.85f)
        }
        if (speed > 0.7f) {
            env.spark(particle.position, min(2.4f, speed), CANONICAL_FORCE_COLORS["wall"]) // canon spark tint
            particle.heat = maxOf(particle.heat, min(0.85f, speed * 0.4f))
        }
    }
}

/** §6.9 — captures matter (held, conserved), then releases on saturation. */
class SinkForce : Force {
    override val token = "sink"
    override val label = "Sink"

    override fun apply(body: Body, particle: Particle, env: Env) {
        if (particle.cap != null || env.dist >= body.absorbR) return
        particle.cap = body
        body.accreted += 1f
        if (body.accreted >= body.capacity) env.supernova(body)
    }
}

/** The canonical nine, in spec order — mirror of Swift's `coreForces()`. */
fun coreForces(): List<Force> = listOf(
    AttractForce(), JetForce(), TetherForce(), WallForce(), StreamForce(),
    RepelForce(), ViscosityForce(), SwirlForce(), SinkForce(),
)
